docs/integration_testing.md

```diff
@@ -0,0 +1,124 @@
+# Writing Integration Tests
+
+## Definition
+
+While _unit tests_ test one single class by stubbing or mocking dependencies, and an end-2-end test relies on the _
+entire_
+system to be present, the _integration test_ tests one particular aspect of a software, which may involve external
+systems.
+
+## TL;DR
+
+Use integration tests only when necessary, keep them concise, implement them in a defensive manner using timeouts and
+randomized names, setup external systems during the workflow.
+
+## When to use them
+
+Generally we should aim at writing unit tests rather than integration tests, because they are simpler, more stable and
+typically run faster. Sometimes that's not (easily) possible, especially when an implementation relies on an external
+system that is not easily mocked or stubbed.
+
+Therefore, in many cases writing unit tests is more involved that writing an integration test, for example say we wanted
+to test our implementation of a CosmosDB-backed queue. We would have to mock the behaviour of the CosmosDB API, which -
+while certainly possible - can get complicated pretty quickly. Now we still might do that for simpler scenarios, but
+eventually we might want to write an integration test that uses a CosmosDB test instance.
+
+## Coding Guidelines
+
+An integration test is annotated with `@IntegrationTest`, which causes the test runner to ignore it unless the
+`RUN_INTEGRATION_TEST` environment variable is set to `true`.
+
+All integration tests should have the `"...IntegrationTest"` postfix to distinguish them clearly from unit tests. They
+should not reside in a separate package, because all tests should maintain package consistency to their test subject.
+
+Any credentials, secrets, passwords, etc. required by the integration tests should be passed in using environment
+variables. A good way to access them is `ConfigurationFunctions#propOrEnv()` because then the credentials can also be
+supplied via system properties.
+
+There is no one-size-fits-all guideline whether to perform setup tasks in the `@BeforeAll` or `@BeforeEach`, it will
+depend on the concrete system you're using. In most cases it is **not** advisable to deploy/provision the external
+system itself in either of them. For example, provisioning a CosmosDB or spinning up a Postgres docker container
```

Comment:
So we're discouraging the use of testing frameworks like testcontainers? I used it a lot in the past and my 2 cents are that it helps a lot the execution of integration tests making them more isolated and controllable